constraint/pkg/client/client_test.go

```diff
@@ -1039,6 +1041,261 @@ func TestClient_AddConstraint(t *testing.T) {
 	}
 }
 
+// TestClient_AddConstraint_withDefaultParams makes sure that we can inject
+// OpenAPIV3 Default parameters into a constraint that does not define its parameters
+// but its template defines defaults for the parameters in the spec's schema.
+func TestClient_AddConstraint_withDefaultParams(t *testing.T) {
+	defaults := []interface{}{"foo", "bar", "baz"}
+	defaults2 := []interface{}{"a", "b", "c"}
+	defaultsJSON := apiextensions.JSON(defaults)
+	defaults2JSON := apiextensions.JSON(defaults2)
+	defaultBuzzJSON := apiextensions.JSON("NOTbuzz")
+
+	target := &handlertest.Handler{}
+	template := cts.New()
+	constraint := cts.MakeConstraint(t, "Fakes", "fakes")
+
+	regoDriver, err := rego.New()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	defaultHandled := map[string]bool{handlertest.TargetName: true}
+	tcs := []struct {
+		name string
+		// validationSpec will be used to mutate the CRD Spec to show different cases.
+		validationSpec *templates.Validation
+		// constraintSpec will be used to simulate the spec for a constraint
+		// like for a constraint that actually defines a parameter.
+		constraintSpec map[string]interface{}
+
+		wantSpec               map[string]interface{}
+		wantHandled            map[string]bool
+		wantAddConstraintError error
+		wantGetConstraintError error
+
+		// driver is pulled out to test actual driver implementations.
+		// defulat is fake.
+		driver drivers.Driver
+	}{
+		// this no op test case is tested implicitly in all
+		// AddConstraint calls outside of this test
+		{
+			name:   "no defaults",
+			driver: regoDriver,
+		},
+		{
+			name: "defaults for one",
+			validationSpec: &templates.Validation{
+				OpenAPIV3Schema: &apiextensions.JSONSchemaProps{
+					Type: "object",
+					Properties: map[string]apiextensions.JSONSchemaProps{
+						"foo": {
+							Type:    "array",
+							Default: &defaultsJSON,
+							Items: &apiextensions.JSONSchemaPropsOrArray{
+								Schema: &apiextensions.JSONSchemaProps{
+									Type: "string",
+								},
+							},
+						},
+					},
+				},
+			},
+			wantSpec: map[string]interface{}{
+				"parameters": map[string]interface{}{
+					"foo": defaults,
+				},
+			},
+			driver: regoDriver,
```

Comment:
Why do we need to test this with anything other than the fake driver?

For ease of writing tests, we should be moving away from non-trivial drivers. This is especially true for this test, since the driver is not part of the relevant functionality.

Reply:
I think it's worth having this functional test to prove that constraints with defaulted params are well handled in the rego driver.

All other cases are using the fake driver to test the flows.

Reply:
We should use the fake driver only. This is not an integration test and furthermore the driver is never actually invoked. Finally, we should not be setting up an expectation that future drivers also need tests here (e.g. K8s admission driver).

Reply:
sounds good, i removed the rego driver from this test func; using fake driver only.